qcfractal/storage_sockets/models/collections_models.py

```diff
@@ -0,0 +1,237 @@
+import datetime
+from sqlalchemy import (Column, Integer, String, DateTime, Boolean, ForeignKey, JSON,
+                        Enum, Float, Binary, Table, ARRAY,
+                        inspect, Index, UniqueConstraint)
+from sqlalchemy.orm import relationship, column_property
+from sqlalchemy import select, func, tuple_, text, cast
+from qcfractal.storage_sockets.models import Base
+from qcfractal.storage_sockets.models import MoleculeORM
+from sqlalchemy.dialects.postgresql import array_agg
+from sqlalchemy.dialects import postgresql
+from sqlalchemy.sql.functions import GenericFunction
+from sqlalchemy.ext.hybrid import hybrid_property
+
+
+# class json_agg(GenericFunction):
+#     type = postgresql.JSON
+
+class json_build_object(GenericFunction):
+    type = postgresql.JSON
+
+
+class CollectionORM(Base):
+    """
+        A base collection class of precomuted workflows such as datasets, ..
+
+        This is a dynamic document, so it will accept any number of
+        extra fields (expandable and uncontrolled schema)
+    """
+
+    __tablename__ = "collection"
+
+    id = Column(Integer, primary_key=True)
+    collection_type = Column(String)  # for inheritance
+
+    collection = Column(String(100), nullable=False)
+    lname = Column(String(100), nullable=False)
+    name = Column(String(100), nullable=False)
+
+    tags = Column(JSON)
+    tagline = Column(String)
+    extra = Column(JSON)  # extra data related to specific collection type
+
+    def update_relations(self, **kwarg):
+        pass
+
+
+    __table_args__ = (
+        Index('ix_collection_lname', "collection", "lname", unique=True),
+        Index('ix_collection_type', 'collection_type'),
+    )
+
+    __mapper_args__ = {'polymorphic_on': 'collection_type'}
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+class DatasetMixin:
+    """
+        Mixin class for common Dataset attributes.
+    """
+
+    default_benchmark = Column(String, nullable=True)
+    default_keywords  = Column(JSON, nullable=True)
+
+    default_driver = Column(String, nullable=True)
+    default_units = Column(String, nullable=True)
+    alias_keywords = Column(JSON, nullable=True)
+    default_program  = Column(String, nullable=True)
+
+    contributed_values = Column(JSON)
+
+    provenance = Column(JSON)  # TODO: in extra?
+
+    history_keys = Column(ARRAY(String, as_tuple=True))
+    history = Column(ARRAY(String, as_tuple=True))
+
+
+class DatasetRecordsORM(Base):
```

Comment:
Just checking, are these concrete inherited tables? Only reason I ask is all collections should share an Id.

Reply:
No, they are just like procedures, they share ids